### ISSUE TITLE
Add takeuntil to signals/processes

### ISSUE THREAD

**Issue by pmros** [joaomilho/act#18]
It would be nice if you add `takeuntil` to signals/proceses, like in rxjs, for example. 
I need `takeuntil` to manage drag and drop (svg has no drag and drop events support). The pseudo-code could be something like:
```
mousedown.flatmap(mousemove).takeuntil(mouseup)
```
I think it should be easy to code but I don't know much about signals/observables/streams implementation.
Thank you.

**Comment by joaomilho** [comment]
Sure.

**Comment by joaomilho** [comment]
@pmros do you have a more complete example of drag & drop w/ Rx? I've found http://theburningmonk.com/2011/03/drag-and-drop-using-reactive-extensions-for-javascript/

I guess the two goals here are a) to have a lower level infrastructure for this sort of use case and b) to have higher level API for the most common use cases, like this https://github.com/jvoigtlaender/elm-drag

It would be nice to have an API like:

```
const svg = ['svg', [
  ['circle', {...track}]
]]
```

And track would have a similar interface, emitting `lift`, `moveBy` and `release` events.

**Comment by pmros** [comment]
It look good for me. But the first thing is create more signals/processes (i.e. `takeuntil`), then create new events (i.e. `dragstart`) upon simple events (i.e. `mousedown`, `mousemove`, `mousedown`) and finally create an optional high-level events upon low-level events. I think we need a good low-level (signals) support before we should think about higher abstractions.
The drag and drop example you link is fine. Actually it's really short and simple, if you ignore all the coordinates/offset stuff.
Any idea about `takeuntil` implementation?

**Comment by pmros** [comment]
I've been playing with Act and I've notice how easy is to create new signal processes! 
`takeUntil` signa process could be something like:
```javascript
const takeUntil = curry((esStop, es, next) => {
  let stop = false
  esStop(() => stop = true)
  es((value) => { if(!stop) next(value) })
})
```
For example, you can use it to create a custom subscription called positionUntilClick (something like positionThrottled from subscriptions/mouse). 
I'll add a pull request for takeUntil when I've created a test for it. Then you could close this issue. Later you could add higher level abstractions (like track).

Another useful signal process would be something like `takeBetween`. Pseudo-code for drag and drop:
```
takeBeetween(merge(mousedown, mouseup), mousemove)
```
 I'll try to add another pull request for `takeBeetween` too.

Thank you again.

**Comment by pmros** [comment]
My drag and drop example:
```javascript
const view = (model) =>
  ['svg', {height: 1000, width: 1000}, [
    ['circle', {cx: model.position[0], cy: model.position[1], r: 40}]
  ]]

const mouseDown = fromEvent(window, 'mousedown').start()
const mouseUp = fromEvent(window, 'mouseup').start()

const takeBetween = curry((es1, es2, next) => {
  let on = false
  es1(() => on = !on)
  es2((value) => { if(on) next(value) })
})

const subscriptions = {
  position: takeBetween(merge(mouseDown, mouseUp))(position)
}

const initialState = { position: [100, 100] }
const reducer = (state = initialState, { type, payload }) => (
  { ...state, [type]: payload }
)

main(view, { subscriptions, reducer })
```
It works pretty well. No offset stuff included.
I'll add a `takeBeetween` pull request (like `takeUntil` one).

The  more I play with Act, the more I love it! :heart:

**Comment by joaomilho** [comment]
Wow, I put some hours on this, but had to get back to work (I'm working on a presentation for CurryOn and doing exchange at the Stockholm office), and the solution looks great. It's simple and easy to understand. Great job!

I'll get back to act fully around day 20.

**Comment by joaomilho** [comment]
Solved by #25
